doc/developer/design/20230802_explain_running_dataflows.md

```diff
@@ -0,0 +1,262 @@
+    duration,            // Always use `Duration::default()` here.
+};
+
+// Constructed using values from a `CatalogItem` field.
+// Use `format` from the enclosing `EXPLAIN` plan.
+Explainable::new(&mut plan).explain(format, context)
+```
+
+Basically
+
+- The `config` is something only depends on the current catalog state.
+- The `humanizer` value can be the current catalog state because item IDs are stable and cannot be recycled.
+- The `finishing` and `duration` values an be constants.
+
+so the only time-dependent information that we need to record is the `plan` and the `used_indexes`.
+The next section expands how we plan to do that with more concrete implementation details.
+
+# Reference explanation
+[reference-explanation]: #reference-explanation
+
+## Extend `CatalogItem` variants
+
+Extend and make an extra field in `Index` and `MaterializedView` with the following fields:
+
+```rust
+// idential extensions to `index`
+struct MaterializedView {
+    ...
+    optimized_plan: Box<mz_expr::MirRelationExpr>,
+    physical_plan: Box<mz_compute_client::Plan>,
+    used_indexes: UsedIndexes,
+    filter_pushdown: bool,
+}
+```
+
+## Change `CatalogItem` constructors
+
+Change client code that creates new `Index` and `MaterializedView` in order to complete the entire optimization path upfront and obtain values for the new fields to the `CatalogItem` instance.
+
+## Add new `DataflowDescription` constructors
+
+The `plan` value in
+
+```rust
+Explainable::new(&mut plan).explain(format, context)
+```
+
+is actually of one of the following two types:
+
+- `DataflowDescription<OptimizedMirRelationExpr>` for `EXPLAIN OPTIMIZED PLAN`
+- `DataflowDescription<Plan>` for `EXPLAIN PHYSICAL PLAN`
+
+However, the only mechanism to construct a `DataflowDescription` that we have at the moment is to start from the previous optimization stage.
+We therefore have to define `DataflowDescription` constructors that reflect the state of a `DataflowDescription` at the end of the `OPTIMIZED PLAN` and `PHYSICAL PLAN` stages.
+The new `sequence_explain_~` methods will then use these constructors to create a `plan` instance from the extended `CatalogState`.
+
+## Refactor `EXPLAIN` handling
+
+Add new `sequence_explain_index` and `sequence_explain_materialized_view` methods. The workflow in these methods should be roughly as follows:
+
+1. Grab a `CatalogItem` for the explained object from the `CatalogState`.
+2. Create an `ExplainContext` using the current context and the `ExplainInfo` attached to the obtained `CatalogItem`.
+3. Create a `DataflowDescription` from the current context and the `ExplainInfo` attached to the obtained `CatalogItem`.
+4. Call `Explainable::new(&mut plan).explain(format, context)`.
+
+
+# Rollout
+[rollout]: #rollout
+
+The proposed implementation path allows to stage the rollout by first merging the necessary changes to the `CatalogItem` variants and observing how the additional data affects our existing environments.
```

Comment:
IMO: Don't worry about doing this in multiple stages. We already have lots of other stuff that uses more memory.

Reply:
OK!